internal/fields/dependency_manager.go

```diff
@@ -0,0 +1,213 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package fields
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+
+	"github.com/elastic/elastic-package/internal/common"
+	"github.com/elastic/elastic-package/internal/configuration/locations"
+	"github.com/elastic/elastic-package/internal/logger"
+	"github.com/elastic/elastic-package/internal/packages/buildmanifest"
+)
+
+const (
+	ecsSchemaName      = "ecs"
+	gitReferencePrefix = "git@"
+
+	ecsSchemaFile = "fields.ecs.yml"
+	ecsSchemaURL  = "https://raw.githubusercontent.com/elastic/ecs/%s/generated/beats/%s"
+)
+
+// DependencyManager is responsible for resolving external field dependencies.
+type DependencyManager struct {
+	schema map[string][]FieldDefinition
+}
+
+// CreateFieldDependencyManager function creates a new instance of the DependencyManager.
+func CreateFieldDependencyManager(deps buildmanifest.Dependencies) (*DependencyManager, error) {
+	schema, err := buildFieldsSchema(deps)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't build fields schema")
+	}
+	return &DependencyManager{
+		schema: schema,
+	}, nil
+}
+
+func buildFieldsSchema(deps buildmanifest.Dependencies) (map[string][]FieldDefinition, error) {
+	schema := map[string][]FieldDefinition{}
+	ecsSchema, err := loadECSFieldsSchema(deps.ECS)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't load fields")
+	}
+	schema[ecsSchemaName] = ecsSchema
+	return schema, nil
+}
+
+func loadECSFieldsSchema(dep buildmanifest.ECSDependency) ([]FieldDefinition, error) {
+	if dep.Reference == "" {
+		logger.Debugf("ECS dependency isn't defined")
+		return nil, nil
+	}
+
+	gitReference, err := asGitReference(dep.Reference)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't process the value as Git reference")
+	}
+
+	loc, err := locations.NewLocationManager()
+	if err != nil {
+		return nil, errors.Wrap(err, "error fetching profile path")
+	}
+
+	cachedSchemaPath := filepath.Join(loc.FieldsCacheDir(), ecsSchemaName, gitReference, ecsSchemaFile)
+	content, err := ioutil.ReadFile(cachedSchemaPath)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return nil, errors.Wrapf(err, "can't read cached schema (path: %s)", cachedSchemaPath)
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		logger.Debugf("Pulling ECS dependency using reference: %s", dep.Reference)
+
+		url := fmt.Sprintf(ecsSchemaURL, gitReference, ecsSchemaFile)
+		logger.Debugf("Schema URL: %s", url)
+		resp, err := http.Get(url)
+		if err != nil {
+			return nil, errors.Wrapf(err, "can't download the online schema (URL: %s)", url)
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != 200 {
+			return nil, fmt.Errorf("unexpected HTTP status code: %d", resp.StatusCode)
```

Comment:
Error is a bit confusing if the version specified doesn't exist:
```
Error: updating files failed: updating readme file README.md failed: rendering Readme failed: executing template failed: template: README.md:3:2: executing "README.md" at <fields "access">: error calling fields: can't create fields validator instance (path: /home/jaime/gocode/src/github.com/elastic/elastic-package/test/packages/nginx/data_stream/access): can't create field dependency manager: can't build fields schema: can't load fields: unexpected HTTP status code: 404
```

It'd be great if somehow it could directly say that the reference of ecs defined in `build.yml` doesn't exist.

Reply:
I adjusted the error message, but I'm not sure about passing this error up. It goes through internals of `text/template` and it helps with figuring out the execution path.

If you think that it will improve user experience, I can introduce additional code to verify dependencies first.

Reply:
Thanks, we can go on like this by now and reconsider later if needed.